src/tests/ftest/scripts/main.sh

```diff
@@ -220,6 +220,10 @@ if [[ "${TEST_TAG_ARG}" =~ soak ]]; then
     fi
 fi
 
+if [[ "${TEST_TAG_ARG}" =~ memcheck ]]; then
+  export WITH_VALGRIND="memcheck"
+fi
+
```

Comment:
This seems to be overriding the value of `$WITH_VALGRIND` that https://github.com/daos-stack/pipeline-lib/blob/027ee8f28079b445d9b9751b2352be6116fb3425/vars/runTestFunctionalV2.groovy#L58 is passing in.

It feels like the value being passed in should be canonical rather than deriving it based on test tags.

Reply:
I'm not sure I follow, but the `$WITH_VALGRIND` usage does seem to be a little out of step with the way our Jenkins/avocado setup works.  It needs to override `$WITH_VALGRIND` on a per-tag basis because it runs cart tests in two modes: with and without valgrind.  If not via the avocado tag, where should `$WITH_VALGRIND` get set?

Reply:
> I'm not sure I follow, but the `$WITH_VALGRIND` usage does seem to be a little out of step with the way our Jenkins/avocado setup works. It needs to override `$WITH_VALGRIND` on a per-tag basis because it runs cart tests in two modes: with and without valgrind.

But those two modes are in two different stages, yes?  The stage (Functional on CentOS 7 with Valgrind) that wants to run them with Valgrind does so through https://github.com/daos-stack/pipeline-lib/blob/027ee8f28079b445d9b9751b2352be6116fb3425/vars/runTestFunctionalV2.groovy#L58 setting `$WITH_VALGRIND` for the stage.  In other stages, that value will not be set because `stage_info.get('with_valgrind', '')` won't set it, right?

> If not via the avocado tag, where should `$WITH_VALGRIND` get set?

The confusion is that you are setting it's value [here](https://github.com/daos-stack/pipeline-lib/blob/027ee8f28079b445d9b9751b2352be6116fb3425/vars/runTestFunctionalV2.groovy#L58) initially and then overriding it on lines 223-225 of this PR.  It seems that [the caller setting it](https://github.com/daos-stack/pipeline-lib/blob/027ee8f28079b445d9b9751b2352be6116fb3425/vars/runTestFunctionalV2.groovy#L58) should be sufficient for the stage, no?

Reply:
> > I'm not sure I follow, but the `$WITH_VALGRIND` usage does seem to be a little out of step with the way our Jenkins/avocado setup works. It needs to override `$WITH_VALGRIND` on a per-tag basis because it runs cart tests in two modes: with and without valgrind.
> 
> But those two modes are in two different stages, yes? 

Yes.

> The stage (Functional on CentOS 7 with Valgrind) that wants to run them with Valgrind does so through https://github.com/daos-stack/pipeline-lib/blob/027ee8f28079b445d9b9751b2352be6116fb3425/vars/runTestFunctionalV2.groovy#L58 setting `$WITH_VALGRIND` for the stage. In other stages, that value will not be set because `stage_info.get('with_valgrind', '')` won't set it, right?
 
Right.

> > If not via the avocado tag, where should `$WITH_VALGRIND` get set?
> 
> The confusion is that you are setting it's value [here](https://github.com/daos-stack/pipeline-lib/blob/027ee8f28079b445d9b9751b2352be6116fb3425/vars/runTestFunctionalV2.groovy#L58) initially and then overriding it on lines 223-225 of this PR. It seems that [the caller setting it](https://github.com/daos-stack/pipeline-lib/blob/027ee8f28079b445d9b9751b2352be6116fb3425/vars/runTestFunctionalV2.groovy#L58) should be sufficient for the stage, no?

I see what you mean now, but I believe the runTestFunctional change is the unnecessary code not doing anything, which is why [line 223](https://github.com/daos-stack/daos/pull/4712/files#diff-b57aa9d6a26c98fa1cc527d81bd155dbc6ac3dc2db18f851296229697906645fR223) was added to this PR.

Reply:
> I see what you mean now, but I believe the runTestFunctional change is the unnecessary code not doing anything,

It would be doing something if it was not getting overridden in this PR though, I believe.  And doing it based on the stage name feels like the right solution, not inferring it based on tag names since it's the point of the **stage** to run tests with valgrind enabled.

> which is why [line 223](https://github.com/daos-stack/daos/pull/4712/files#diff-b57aa9d6a26c98fa1cc527d81bd155dbc6ac3dc2db18f851296229697906645fR223) was added to this PR. 

Understood.  But that addition seems redundant when https://github.com/daos-stack/pipeline-lib/blob/027ee8f28079b445d9b9751b2352be6116fb3425/vars/runTestFunctionalV2.groovy#L58 will just set `$WITH_VALGRIND=memcheck` based on the stage info, because it's the **stage** that is driving the enablement of valgrind, not test tags.

So my suggestion is to revert https://github.com/daos-stack/pipeline-lib/pull/216/commits/ddad8212f0df71525faba88bc893f5fabbdc8351 and remove https://github.com/daos-stack/daos/blob/4a3fae9b4282caee5efdd6edb48d2d00f6d21457/src/tests/ftest/scripts/main.sh#L223-L226 and confirm that you are getting valgrind results from the stage, which by my reading of the code, you should be.  If you are not, then it should not be difficult to determine why and rectify it.

Reply:
> > I see what you mean now, but I believe the runTestFunctional change is the unnecessary code not doing anything,
> 
> It would be doing something if it was not getting overridden in this PR though, I believe. And doing it based on the stage name feels like the right solution, not inferring it based on tag names since it's the point of the **stage** to run tests with valgrind enabled.
> 
> > which is why [line 223](https://github.com/daos-stack/daos/pull/4712/files#diff-b57aa9d6a26c98fa1cc527d81bd155dbc6ac3dc2db18f851296229697906645fR223) was added to this PR.
> 
> Understood. But that addition seems redundant when https://github.com/daos-stack/pipeline-lib/blob/027ee8f28079b445d9b9751b2352be6116fb3425/vars/runTestFunctionalV2.groovy#L58 will just set `$WITH_VALGRIND=memcheck` based on the stage info, because it's the **stage** that is driving the enablement of valgrind, not test tags.
> 
> So my suggestion is to revert [daos-stack/pipeline-lib@ddad821](https://github.com/daos-stack/pipeline-lib/commit/ddad8212f0df71525faba88bc893f5fabbdc8351) and remove
> 
> https://github.com/daos-stack/daos/blob/4a3fae9b4282caee5efdd6edb48d2d00f6d21457/src/tests/ftest/scripts/main.sh#L223-L226
> 
> and confirm that you are getting valgrind results from the stage, which by my reading of the code, you should be. If you are not, then it should not be difficult to determine why and rectify it.

IIRC, the `WITH_VALGRIND` env var was getting unset in the chain of scripts, e.g., Jenkins => `test_main.sh` => `ftest.sh`.  I've made the suggested changes, but there's another blocker: Functional tests are broken on this branch, and I don't know the cause (see [this comment](https://github.com/daos-stack/pipeline-lib/pull/216#issuecomment-856926350)).

Reply:
See https://github.com/daos-stack/pipeline-lib/pull/216#issuecomment-857066617 for the Functional test issue.

Reply:
> ... and confirm that you are getting valgrind results from the stage, which by my reading of the code, you should be. If you are not, then it should not be difficult to determine why and rectify it.

K, `launch.py` is working again:  

  https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-4712/65/

But somewhere in our chain of scripts, `WITH_VALGRIND` seems to unset:
   Jenkins runner => `test_main.sh` => `ftest.sh` => `launch.py`

Do we need `AcceptEnv WITH_VALGRIND` in [`sshd_config`](https://linux.die.net/man/5/sshd_config)?